### PR TITLE
Feature : Alias show - Alias edit

### PR DESCRIPTION
**Pull request by tugbataluy**
What This PR Does

This PR adds two new commands to improve the management of client command aliases:

- lxc alias show: Displays all current aliases in a structured YAML format to stdout.
- lxc alias edit: Allows bulk, simultaneous editing of all aliases. It supports interactive editing via the system editor and piping of configuration via stdin.

 lxc alias edit: Allows bulk, simultaneous editing of all aliases. It supports interactive editing via the system editor and piping of configuration via stdin.

Why This Is Needed

These commands replace the inefficient process of adding or removing aliases one-by-one, enabling quick auditing and configuration deployment through scripting or a single editing session.

Implementation Highlights

-     Consistency: Both commands align with standard LXC configuration commands (like lxc profile edit)
-     Safety: lxc alias edit includes a required validation loop for interactive edits to prevent saving invalid YAML syntax.
-     Compatibility: This implementation fully relies on the existing lxc alias commands and c.global.configPath for persistence in the standard YAML configuration file.

Example Usage
 ```
 lxc alias show
 lxc alias edit
 cat aliases.yaml | lxc alias edit 
```

### CODE REVIEW

File: lxc/alias.go [hunk 1]
```diff
@@ -245,3 +259,194 @@ func (c *cmdAliasRemove) run(cmd *cobra.Command, args []string) error {
...
+	alias  *cmdAlias
+}
+
+// Command creates a Cobra command to show all aliases in YAML format.
+func (c *cmdAliasShow) command() *cobra.Command {
+	cmd := &cobra.Command{}
+	cmd.Use = usage("show")
+	cmd.Short = i18n.G("Show aliases in YAML format")
+	cmd.Long = cli.FormatSection(i18n.G("Description"), i18n.G(
+		`Show aliases im YAML format`))
```
Comment (minaelee) [inline lxc/alias.go]:
```suggestion
		`Show aliases in YAML format`))
```

Comment (tugbataluy) [reply lxc/alias.go]: Thanks Minae, I fixed it :)

**Comment by minaelee** [comment]
I realized there is also a how-to for `lxc alias` that should be updated: https://documentation.ubuntu.com/lxd/latest/howto/lxc_alias/

It's in doc > howto > lxc_alias.md. Let me know if you'd like any guidance. Ty!

**Comment by tugbataluy** [comment]
> I realized there is also a how-to for `lxc alias` that should be updated: https://documentation.ubuntu.com/lxd/latest/howto/lxc_alias/
> 
> It's in doc > howto > lxc_alias.md. Let me know if you'd like any guidance. Ty!

Hi @minaelee,
I added the edit and show commands to the document you mentioned. Thank you.

File: lxc/alias.go [hunk 2]
```diff
@@ -245,3 +258,166 @@ func (c *cmdAliasRemove) run(cmd *cobra.Command, args []string) error {
...
+}
+
+// Command creates a Cobra command to edit aliases either via interactive editor or via pipe to stdin.
+func (c *cmdAliasEdit) command() *cobra.Command {
+	cmd := &cobra.Command{}
+	cmd.Use = usage("edit")
+	cmd.Short = i18n.G("Edit aliases")
+
+	cmd.Example = cli.FormatSection("", i18n.G(`lxc alias edit
+	Update the aliases via interactive terminal.
```
Comment (MusicDin) [inline lxc/alias.go]: Should we use `Edit` instead of `Update`?

File: lxc/alias.go [hunk 3]
```diff
@@ -245,3 +258,166 @@ func (c *cmdAliasRemove) run(cmd *cobra.Command, args []string) error {
...
+}
+
+// HelpTemplate returns a sample YAML representation of aliases and guidelines for editing.
+func (c *cmdAliasEdit) helpTemplate() string {
+	return i18n.G(
+		`### This is a YAML representation of the aliases.
+### Any line starting with a '#' will be ignored.
+###
+### A sample aliases configuration looks like:
+### list: "list -c ns46S"
```
Comment (MusicDin) [inline lxc/alias.go:339]: Are aliases displayed with `aliases` namespace? If so, maybe we can include the namespace in the example?

```yaml
### aliases:
###   list: "list -c ns46S"
###   ...
```

File: lxc/alias.go [hunk 4]
```diff
@@ -245,3 +258,166 @@ func (c *cmdAliasRemove) run(cmd *cobra.Command, args []string) error {
...
+### my-list: "list -c ns46S"
+### start-all: "start --all"
+###
+### Note that aliases are simple key-value pairs.`)
```
Comment (MusicDin) [inline lxc/alias.go]: Can we drop `simple`? They are just key-value pairs :)

File: lxc/alias.go [hunk 5]
```diff
@@ -245,3 +258,166 @@ func (c *cmdAliasRemove) run(cmd *cobra.Command, args []string) error {
...
+		}
+
+		err = yaml.Unmarshal(contents, &config)
+		if err != nil {
+			return nil
+		}
+
+		// Update aliases and save config
+		conf.Aliases = newAliases
+		importedCount := len(newAliases)
```
Comment (MusicDin) [inline lxc/alias.go]: Do we need to verify new aliases? Maybe at least to ensure key/value are not empty?

EDIT:
To clarify, I just want to make sure we do not allow storing something that could cause a parsing error later. Such error would also affect parsing remotes which happens for almost any command.

However, such error should be caught during marshaling.

File: lxc/alias.go [hunk 6]
```diff
@@ -245,3 +258,166 @@ func (c *cmdAliasRemove) run(cmd *cobra.Command, args []string) error {
...
+	}
+
+	for {
+		// Parse the text received from the editor
+		newAliases := make(map[string]string)
+		err = yaml.Unmarshal(content, &newAliases)
+		if err == nil {
+			// Update aliases and save config
+			conf.Aliases = newAliases
+			err = conf.SaveConfig(c.global.confPath)
```
Comment (MusicDin) [inline lxc/alias.go]: I see you first check if the parsing error is nil. In such case you try to store the config with new aliases. If storing fails, you reopen the terminal.

I'm wondering what is the error if storing a config fails? If we cannot save the config, is there a good reason to offer user to edit configuration again?
Maybe you can just error out in such case:

```go
for {
		// Parse the text received from the editor
		newAliases := make(map[string]string)
		err = yaml.Unmarshal(content, &newAliases)
		if err != nil {
		    // Respawn the editor if there was an error
			fmt.Fprintf(os.Stderr, i18n.G("Alias parsing error: %s")+"\n", err)
			fmt.Println(i18n.G("Press enter to open the editor again or ctrl+c to abort change"))
			_, err := os.Stdin.Read(make([]byte, 1))
			if err != nil {
				return err
			}

			content, err = shared.TextEditor("", content)
			if err != nil {
				return err
			}

			continue
		}

	    // Update aliases and save config
	    conf.Aliases = newAliases
	    return conf.SaveConfig(c.global.confPath)
	}
```

File: lxc/alias.go (hunk 3 above)
Comment (tugbataluy) [reply lxc/alias.go:339]: It depends on the command if user uses lxc alias list it only shows key-value pairs. If user uses lxc alias show it looks like:
aliases:
  details: config show
  list: list -c ns4
  list_network: network list
  list_storage: storage list
But  interactive edit basically shows key-value pairs as lxc alias list does.

Comment (MusicDin) [reply lxc/alias.go:339]: > But interactive edit basically shows key-value pairs as lxc alias list does.

Perfect, thanks 👍

**Comment by tomponline** [comment]
Moving back to draft after our meeting, ta

File: lxc/alias.go (hunk 3 above)
Comment (tugbataluy) [reply lxc/alias.go:339]: Now both lxc alias list -f yaml's and lxc alias show output's look like:
details: config show
list: list -c ns4
list_network: network list
list_storage: storage list

File: lxc/alias.go (hunk 5 above)
Comment (tugbataluy) [reply lxc/alias.go]: When user edits aliases interactively user can see the details of the key-value pairs, but when edits non-interactively I think we should give information to the user (at least that is what I was trying to do)

File: lxc/alias.go (hunk 3 above)
Comment (tomponline) [reply lxc/alias.go:339]: Thanks @tugbataluy !

**Comment by simondeziel** [comment]
> Copilot encountered an error and was unable to review this pull request. You can try again by re-requesting a review.

I wonder if Copilot has issue when there are too many files to review which is probably worsen by updating the translation. Maybe it's worth dropping the translation changes to try if it works better?

Or maybe it's due to GH partial outage that is ongoing.

**Comment by MusicDin** [comment]
> I wonder if Copilot has issue when there are too many files to review which is probably worsen by updating the translation. Maybe it's worth dropping the translation changes to try if it works better?

That's a good point. On my PRs where translations are updated, Copilot also does not work.

File: lxc/alias.go [hunk 7]
```diff
@@ -245,3 +258,154 @@ func (c *cmdAliasRemove) run(cmd *cobra.Command, args []string) error {
...
+		newAliases := make(map[string]string)
+		config := struct {
+			Aliases map[string]string `yaml:"aliases"`
+		}{
+			Aliases: newAliases,
+		}
+
+		err = yaml.Unmarshal(contents, &config)
+		if err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (tugbataluy) [reply lxc/alias.go]: Done

File: lxc/alias.go [hunk 8]
```diff
@@ -245,3 +258,154 @@ func (c *cmdAliasRemove) run(cmd *cobra.Command, args []string) error {
...
+		}
+
```
[the remark this answers is not in the document]
Comment (tugbataluy) [reply lxc/alias.go:376]: Done

File: doc/howto/lxc_alias.md [hunk 9]
```diff
@@ -16,4 +16,18 @@ For example, to always ask for confirmation when deleting an instance, create an
     lxc alias add delete "delete -i"
 
 To see all configured aliases, run [`lxc alias list`](lxc_alias_list.md).
+
+To view all aliases in YAML format (useful for exporting or inspection), run [`lxc alias show`](lxc_alias_show.md).
```
Comment (minaelee) [inline doc/howto/lxc_alias.md]:
````suggestion
To [view all aliases in YAML format](lxc_alias_show.md) (useful for exporting or inspection), run:

```bash
lxc alias show
```
````

File: doc/howto/lxc_alias.md [hunk 10]
```diff
@@ -16,4 +16,18 @@ For example, to always ask for confirmation when deleting an instance, create an
...
+To edit all aliases interactively or via file input, run [`lxc alias edit`](lxc_alias_edit.md). This command opens your system's default text editor and allow you to modify all aliases at once. You can also pipe alias configurations to this command:
+
+    # Export aliases to a file
+    lxc alias show > aliases.yaml
+
+    # Import aliases from a file
+    lxc alias edit < aliases.yaml
+
+    # Import from pipe
+    cat aliases.yaml | lxc alias edit
```
Comment (minaelee) [inline doc/howto/lxc_alias.md]:
````suggestion
To [edit all aliases](lxc_alias_edit.md) interactively or via file input, run:

```bash
lxc alias edit
```

This command opens your system's default text editor and allows you to modify all aliases at once.

You can also pipe alias configurations to this command. Examples:

```bash
# Export aliases to a file
lxc alias show > aliases.yaml

# Import aliases from a file
lxc alias edit < aliases.yaml

# Import from pipe
cat aliases.yaml | lxc alias edit
```
````

File: doc/howto/lxc_alias.md (hunk 9 above)
Comment (minaelee) [reply doc/howto/lxc_alias.md]: Prefer command syntax in code blocks rather than inline so that the user can see it clearly when scanning. Prefer backticks for code blocks instead of tabs for clarity and the ability to specify a language for syntax highlighting.

File: lxc/alias.go [hunk 11]
```diff
@@ -245,3 +259,170 @@ func (c *cmdAliasRemove) run(cmd *cobra.Command, args []string) error {
...
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	// Convert aliases to YAML and print
+	data, err := yaml.Marshal(&conf.Aliases)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("%s", data)
```
Comment (tomponline) [inline lxc/alias.go]: We try to avoid using `Printf` et al where possible to avoid reflection.

Here we would use `fmt.Print(string(data))`

Comment (tomponline) [reply lxc/alias.go]: Should this include a new line too, via `fmt.Println()`?

**Review by tomponline — CHANGES_REQUESTED** [review]
Please can tests be added - suggest in a new test function in the main test suite.

Also please can you rebase. Ta

File: lxc/alias.go (hunk 11 above)
Comment (tugbataluy) [reply lxc/alias.go]: OK, thanks.
For fmt.Println() are we talking about the end of the alias list? If that is the case I can use it.

**Comment by tomponline** [comment]
LGTM thanks!

Ive asked @simondeziel to take a look at the tests for consistency. Thanks!

File: test/suites/alias.sh [hunk 12]
```diff
@@ -0,0 +1,158 @@
+test_alias() {
+  ensure_import_testimage
+
+  # Test 1: Add a new alias
+  echo "Test: Add a new alias"
+  lxc alias add my-list "list -c ns46S"
+  lxc alias list --format csv | grep "my-list,list -c ns46S" > /dev/null
```
Comment (simondeziel) [inline test/suites/alias.sh]: Since it's a full line match, `-x` is recommended and `-F` is extra precaution to convey that we are searching for a fixed string (!= regex).
```suggestion
  lxc alias list --format csv | grep -xF "my-list,list -c ns46S"
```

Also, we typically do not suppress the output. It does make CI logs more verbose but it also aid navigating them.

Those suggestions apply to many occurrences below that I won't pin point :)

File: test/suites/alias.sh [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+  # Test 2: Try to add duplicate alias (should fail)
+  echo "Test: Prevent adding duplicate alias"
+  ALIAS_ERR="$(! lxc alias add my-list "list -c -s" 2>&1)"
```
Comment (simondeziel) [inline test/suites/alias.sh]: Just in case it succeeds unexpectedly. Granted, this one is mostly for consistency as you inspect the output right after.
```suggestion
  ALIAS_ERR="$(! lxc alias add my-list "list -c -s" 2>&1 || echo fail)"
```

File: test/suites/alias.sh [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+  echo "${ALIAS_ERR}" | grep "Alias my-list already exists" > /dev/null
```
Comment (simondeziel) [inline test/suites/alias.sh]: Whenever possible, full matches are even better, especially when done with shell comparison.
```suggestion
  [ "${ALIAS_ERR}" = "Error: Alias my-list already exists" ]
```

File: test/suites/alias.sh [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+
+  # Test 3: Show alias in YAML format
+  echo "Test: Show alias in YAML format"
+  lxc alias show | grep "my-list: list -c ns46S" > /dev/null
+
+  # Test 4: Rename an alias
+  echo "Test: Rename an alias"
+  lxc alias rename my-list my-list2
+  ! lxc alias list --format csv | grep "my-list," > /dev/null || false
```
Comment (simondeziel) [inline test/suites/alias.sh]: It's best to be extra defensive when dealing with "user input". `grep "^my-list,"` would make sure that it only ever matches on the first column and not somewhere random in the alias definition itself.

File: test/suites/alias.sh [hunk 16]
```diff
@@ -0,0 +1,158 @@
+test_alias() {
+  ensure_import_testimage
+
+  # Test 1: Add a new alias
+  echo "Test: Add a new alias"
```
Comment (simondeziel) [inline test/suites/alias.sh]: The comment doesn't add much so I'd drop it. It only risks getting out of sync IMHO.

The `echo` however is quite nice to aid navigating CI logs :) Feel free to include the test number in the `echo` if you find it useful.

File: test/suites/alias.sh [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+  lxc alias show | grep "my-list: list -c ns46S" > /dev/null
+
+  # Test 4: Rename an alias
+  echo "Test: Rename an alias"
+  lxc alias rename my-list my-list2
+  ! lxc alias list --format csv | grep "my-list," > /dev/null || false
+  lxc alias list --format csv | grep "my-list2,list -c ns46S" > /dev/null
+
+  # Test 5: Try to rename to existing alias (should fail)
+  echo "Test: Prevent rename to existing alias"
```
Comment (simondeziel) [inline test/suites/alias.sh]: Nice edge case testing :)

File: test/suites/alias.sh [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+  done
+
+  # Check aliases
+  for alias_name in "${!special_aliases[@]}"; do
+    lxc alias list --format csv | grep "${alias_name},${special_aliases[$alias_name]}" > /dev/null
+  done
+
+  # Test 18: Test bulk operations via edit
+  echo "Test: Bulk update aliases via edit"
+  cat <<EOF | lxc alias edit
```
Comment (simondeziel) [inline test/suites/alias.sh]: No need for `cat` for `heredocs`.
```suggestion
  lxc alias edit <<EOF
```

File: test/suites/alias.sh [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+bulk-alias3: "list --format yaml"
+bulk-alias4: "list --format table"
+EOF
+
+  # Check bulk aliases
+  bulk_formats=("csv" "json" "yaml" "table")
+  for i in {1..4}; do
+    format_index=$((i-1))
+    lxc alias list --format csv | grep "bulk-alias${i},list --format ${bulk_formats[$format_index]}" > /dev/null
+  done
```
Comment (simondeziel) [inline test/suites/alias.sh:114]: The hardcoded `{1..4}` part would be best avoided. Since you are already using an array, I'd suggest iterating over the array and compute the index. Gemini proposed this (which I only mildly tweaked):

```
bulk_formats=("csv" "json" "yaml" "table")
for format_index in "${!bulk_formats[@]}"; do
  alias_number=$((format_index + 1))
  format_string=${bulk_formats[format_index]}
  lxc alias list --format csv | grep -xF "bulk-alias${alias_number},list --format ${format_string}"
done
```

File: test/suites/alias.sh [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+  # Verify that Round-trip aliases still exist and being correct
+  for alias_name in "${!roundtrip_aliases[@]}"; do
+    # check with JSON format
+    lxc alias list --format json | jq -e ".[\"${alias_name}\"]" > /dev/null
+
+    # Check with CSV format
+    lxc alias list --format csv | grep "${alias_name},${roundtrip_aliases[$alias_name]}" > /dev/null
+  done
+
+}
```
Comment (simondeziel) [inline test/suites/alias.sh]: Your editor (VSCode does that by default IIRC) is not adding the last `\n` that would avoid this warning.

File: test/suites/alias.sh [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+  # show alias and pipe to edit
+  lxc alias show | lxc alias edit
+
+  # Verify that Round-trip aliases still exist and being correct
+  for alias_name in "${!roundtrip_aliases[@]}"; do
+    # check with JSON format
+    lxc alias list --format json | jq -e ".[\"${alias_name}\"]" > /dev/null
+
+    # Check with CSV format
+    lxc alias list --format csv | grep "${alias_name},${roundtrip_aliases[$alias_name]}" > /dev/null
```
Comment (simondeziel) [inline test/suites/alias.sh]: It's a minor thing but when the output of a given command is relatively short (few 10s of lines), I think it's best to save it in a variable outside of the loop so that it is queried only once.

```
ALIAS_LIST_JSON="$(lxc alias list --format json)"
...
for ...
   jq --exit-status '.["my-list"]' <<< "${ALIAS_LIST_JSON}"
...
```

Also, please use long arg names with `jq`, we are in the process of changing all other occurrences in the test suite :)

File: test/suites/alias.sh (hunk 12 above)
Comment (tugbataluy) [reply test/suites/alias.sh]: In appropriate situations, I used the -xF flag; in situations where I did not perform full line control, I used the -F flag.

File: test/suites/alias.sh (hunk 16 above)
Comment (tugbataluy) [reply test/suites/alias.sh]: I added the numbers to the echo lines

File: test/suites/alias.sh [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+  echo "Test 2: Prevent adding duplicate alias (should fail)"
+  ALIAS_ERR="$(! lxc alias add my-list "list -c -s" 2>&1 || echo fail)"
+  [ "$(echo "${ALIAS_ERR}" | tail -1)" = "Error: Alias my-list already exists" ]
+
+  echo "Test 3: Show alias in YAML format"
+  lxc alias show | grep -xF "my-list: list -c ns46S"
+
+  echo "Test 4: Rename an alias"
+  lxc alias rename my-list my-list2
+  ! lxc alias list --format csv | grep -xF "my-list," || false
```
Comment (simondeziel) [inline test/suites/alias.sh]: `-x` should be dropped for this partial match.

File: test/suites/alias.sh [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+  lxc alias list --format csv | grep -xF "my-list2,list -c ns46S"
+
+  echo "Test 5: Prevent rename to existing alias (should fail)"
+  lxc alias add another-alias "list -c s"
+  ALIAS_ERR="$(! lxc alias rename another-alias my-list2 2>&1 || echo fail)"
+  [ "$(echo "${ALIAS_ERR}" | tail -1)" = "Error: Alias my-list2 already exists" ]
+
+  echo "Test 6: Remove an alias"
+  lxc alias remove my-list2
+  ! lxc alias list --format csv | grep -xF "^my-list2," || false
```
Comment (simondeziel) [inline test/suites/alias.sh]: Same re `-x`

File: test/suites/alias.sh [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+
+  echo "Test 4: Rename an alias"
+  lxc alias rename my-list my-list2
+  ! lxc alias list --format csv | grep -xF "my-list," || false
+  lxc alias list --format csv | grep -xF "my-list2,list -c ns46S"
+
+  echo "Test 5: Prevent rename to existing alias (should fail)"
+  lxc alias add another-alias "list -c s"
+  ALIAS_ERR="$(! lxc alias rename another-alias my-list2 2>&1 || echo fail)"
+  [ "$(echo "${ALIAS_ERR}" | tail -1)" = "Error: Alias my-list2 already exists" ]
```
Comment (simondeziel) [inline test/suites/alias.sh:25]: Oh I didn't know there was a header/line to drop. LGTM as-is but the `grep` also made sense because of that, sorry.

File: test/suites/alias.sh [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+  lxc alias list --format csv | grep -xF "complex,list --format csv --all-projects"
+
+  echo "Test 9: Alias list formats"
+  formats=("json" "yaml" "compact" "table")
+  for format in "${formats[@]}"; do
+    lxc alias list --format "${format}" | grep -F "another-alias"
+  done
+
+  echo "Test 10: Edit aliases via stdin (non-interactive)"
+  cat <<EOF | lxc alias edit
```
Comment (simondeziel) [inline test/suites/alias.sh]: No need for `cat` here.

File: test/suites/alias.sh [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+  lxc alias list --format csv | grep -xF "new-alias1,list --format csv"
+  lxc alias list --format csv | grep -xF "new-alias2,list -c ns"
+
+  echo "Test 11: Prevent clearing all aliases with empty input"
+  ALIAS_ERR="$(! echo "" | lxc alias edit 2>&1 || echo fail)"
+  [ "$(echo "${ALIAS_ERR}" | tail -1)" = "Error: no aliases found in input" ]
+
+  echo "Test 12: Handle invalid YAML in edit (non-interactive)"
+  # Store current aliases count
+  alias_count_before=$(lxc alias list --format json | jq length)
```
Comment (simondeziel) [inline test/suites/alias.sh:59]:
```suggestion
  alias_count_before="$(lxc alias list --format json | jq --exit-status length)"
```

This ensures that `jq` will error our if it receives malformed JSON or cannot compute the length.

New tests should always use `jq --exit-status`. Eventually, there will be a linter script to enforce this.

File: test/suites/alias.sh [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+  ALIAS_ERR="$(! echo "invalid: yaml: [}" | lxc alias edit 2>&1 || echo fail)"
+  echo "${ALIAS_ERR}" | grep -F "yaml:"
+
+  # Verify aliases were not changed
+  alias_count_after=$(lxc alias list --format json | jq length)
+  [ "${alias_count_before}" = "${alias_count_after}" ]
```
Comment (simondeziel) [inline test/suites/alias.sh]:
```suggestion
  lxc alias list --format json | jq --exit-status "length == \"${alias_count_before}\""
```

File: test/suites/alias.sh [hunk 28]
```diff
@@ -0,0 +1,143 @@
...
+
+  echo "Test 13: Verify alias functionality"
+  lxc alias add short-list "list --format csv"
+  lxc init testimage test-alias-instance -d "${SMALL_ROOT_DISK}"
```
Comment (simondeziel) [inline test/suites/alias.sh]: When we can get away with an `--empty` instance, it's always better :)
```suggestion
  lxc init --empty test-alias-instance -d "${SMALL_ROOT_DISK}"
```

File: test/suites/alias.sh [hunk 29]
```diff
@@ -0,0 +1,143 @@
+test_alias() {
+  ensure_import_testimage
```
Comment (simondeziel) [inline test/suites/alias.sh]: This shouldn't be needed with the `--empty` suggestion.

File: test/suites/alias.sh [hunk 30]
```diff
@@ -0,0 +1,143 @@
...
+  lxc alias add short-list "list --format csv"
+  lxc init testimage test-alias-instance -d "${SMALL_ROOT_DISK}"
+  lxc short-list | grep -F "test-alias-instance"
+
+  echo "Test 14: ls alias for list command"
+  lxc alias ls --format csv | grep -F "new-alias1"
+
+  echo "Test 15: rm alias for remove command"
+  lxc alias rm new-alias1
+  ! lxc alias list --format csv | grep -F "^new-alias1," || false
```
Comment (simondeziel) [inline test/suites/alias.sh]: `^` is a regex anchor so it's not compatible with `-F`.

File: test/suites/alias.sh [hunk 31]
```diff
@@ -0,0 +1,143 @@
...
+
+  echo "Test 16: mv alias for rename command"
+  lxc alias mv new-alias2 renamed-alias
+  ! lxc alias list --format csv | grep -F "^new-alias2," || false
```
Comment (simondeziel) [inline test/suites/alias.sh]: Same re `^`

File: test/suites/alias.sh [hunk 32]
```diff
@@ -0,0 +1,143 @@
...
+  # Get alias lists once outside the loop
+  ALIAS_LIST_JSON="$(lxc alias list --format json)"
+  ALIAS_LIST_CSV="$(lxc alias list --format csv)"
+
+  # Verify that Round-trip aliases still exist and being correct
+  for alias_name in "${!roundtrip_aliases[@]}"; do
+    # check with JSON format
+    jq --exit-status '.["'"${alias_name}"'"]' <<< "${ALIAS_LIST_JSON}"
+    # Check with CSV format
+    grep -xF "${alias_name},${roundtrip_aliases[$alias_name]}" <<< "${ALIAS_LIST_CSV}"
```
Comment (simondeziel) [inline test/suites/alias.sh:142]: Since we are using `jq`, might as well also ensure the alias definitions match.
```suggestion
  for alias_name in "${!roundtrip_aliases[@]}"; do
    alias_value="${roundtrip_aliases[$alias_name]}"
    # check with JSON format
    jq --exit-status '.["'"${alias_name}"'"] == "'"${alias_value}"'"' <<< "${ALIAS_LIST_JSON}"
    # Check with CSV format
    grep -xF "${alias_name},${alias_value}" <<< "${ALIAS_LIST_CSV}"
```

File: test/suites/alias.sh [hunk 33]
```diff
@@ -0,0 +1,143 @@
...
+  done
+
```
Comment (simondeziel) [inline test/suites/alias.sh:144]: I believe that aliases will be "left behind" after running this script. I suggest either cleaning them all or maybe it'd be simpler to backup the original `config.yml` and restore it at the end.

File: test/suites/alias.sh (hunk 28 above)
Comment (simondeziel) [reply test/suites/alias.sh]: In fact, I'd avoid creating an instance altogether. Something like that maybe:

```
lxc alias add running 'list -f csv STATUS=RUNNING'
[ "$(lxc running || echo fail)" = "" ]
```

Instances, however small involve quite a few IOs which are best avoided, especially on slow backends like `ceph`.

The other advantage is that you then don't need to `lxc delete test-alias-instance` ;)

File: test/main.sh [hunk 34]
```diff
@@ -658,6 +658,7 @@ if [ "${1:-"all"}" != "snap" ] && [ "${1:-"all"}" != "cluster" ]; then
     run_test test_waitready "waitready"
     run_test test_vm_empty "Empty VM"
     run_test test_vm_pcie_bus "VM PCIe bus numbers"
+    run_test test_alias
```
Comment (simondeziel) [inline test/main.sh]: Could you please make that one of the first tests? This should help surface any missing cleanup bits.

**Comment by tomponline** [comment]
@tugbataluy sorry please can you rebase again (I merged another PR as I saw the tests on this were still running and now it conflicts with the translations). Thanks!

File: lxc/alias.go [hunk 35]
```diff
@@ -245,3 +259,170 @@ func (c *cmdAliasRemove) run(cmd *cobra.Command, args []string) error {
...
+
+	cmd.ValidArgsFunction = func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		if len(args) == 0 {
+			return nil, cobra.ShellCompDirectiveDefault
+		}
+
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	// if stdin isn't a terminal, read text from it
```
Comment (MusicDin) [inline lxc/alias.go]: Not sure if we are still enforcing that, but we tend to start the comments uppercase and end them with a dot.

File: test/suites/alias.sh [hunk 36]
```diff
@@ -0,0 +1,185 @@
+test_alias() {
+  # Get list of original aliases before test
+  local original_alias_list
+  original_alias_list=$(lxc alias list --format json | jq --raw-ouput 'keys[]' || echo "")
+
+  echo "Test 1: Add a new alias"
+  lxc alias add my-list "list -c ns46S"
+  lxc alias list --format csv | grep  -xF "my-list,list -c ns46S"
```
Comment (MusicDin) [inline test/suites/alias.sh]: There are two spaces between `grep` and `-xF`

File: test/main.sh [hunk 37]
```diff
@@ -497,6 +497,7 @@ fi
 
 if [ "${1:-"all"}" != "snap" ] && [ "${1:-"all"}" != "cluster" ]; then
     #run_test test_concurrent "concurrent startup" # Disabled as flaky.
+    run_test test_alias
```
Comment (MusicDin) [inline test/main.sh]: Should there be the name?
Like `run_test test_alias "import/export aliases"`?

File: lxc/alias.go [hunk 38]
```diff
@@ -245,3 +259,170 @@ func (c *cmdAliasRemove) run(cmd *cobra.Command, args []string) error {
...
+		err = yaml.Unmarshal(contents, &newAliases)
+		if err != nil {
+			return err
+		}
+
+		importedCount := len(newAliases)
+		// Prevent clearing all aliases if input is empty
+		if importedCount == 0 {
+			fmt.Fprintln(os.Stderr, i18n.G("No aliases found in input. Aborting to prevent clearing all aliases."))
+			return errors.New("no aliases found in input")
```
Comment (MusicDin) [inline lxc/alias.go]: Error capitalized

Comment (MusicDin) [reply lxc/alias.go]: Also, why is it writing to stderr and returning an error? Will this be shown twice?
